### ISSUE TITLE
Access violation in finalizer of PreparedSqlLiteInsertCommand

### ISSUE THREAD

**Issue by ptsoccer** [praeclarum/sqlite-net#667]
I've been debugging an access violation exception for a couple days in SQL cipher and have come across an interesting issue.

When PreparedSqlLiteInsertCommand gets finalized, it calls SQLite3.Finalize which ends up calling Sqlite3.sqlite3_finalize on the sqlite3_stmt. At the same time, sqlite3_stmt has its own finalizer which also calls Sqlite3.sqlite3_finalize, so a double free ends up happening. I'm not sure why this crash doesn't happen more often, but I suspect the fix is checking to see if disposing is true in the dispose method before calling sqlite3_finalize.

SQLiteConnection also seems to have this issue for the Handle property. It explicitly calls Close2 on Handle if disposing is false, but the sqlite3 class already has a finalizer that will do this.

**Comment by ericsink** [comment]
I don't see how this case can actually happen.  

The finalizer (in the C# sense) for sqlite3_stmt checks to make sure it isn't already done.

**Comment by ptsoccer** [comment]
Only in the dispose method of sqlite3_smt sqlite-net directly calls sqlite3_finalize (https://github.com/praeclarum/sqlite-net/blob/master/src/SQLite.cs#L3022) -> https://github.com/praeclarum/sqlite-net/blob/master/src/SQLite.cs#L3022 -> https://github.com/praeclarum/sqlite-net/blob/master/src/SQLite.cs#L3022 this bypasses the check

**Comment by ptsoccer** [comment]
After reading the code more, I see now what you're saying and now I'm confused as well. The issue is I have a crash dump that seems to indicate this double finalize happened. I will do more research on this.

**Comment by ptsoccer** [comment]
Okay sorry for the stream of thought, but now I know why this can happen. If the sqlite3_stmt finalizer runs and then the PreparedSqlLiteInsertCommand finalizer runs, this can happen. The sqlite3_stmt finalizes and sets _disposed to true, but the PreparedSqlLiteInsertCommand finalizer calls sqlite3_finalize directly which doesn't have the _disposed check.

**Comment by ericsink** [comment]
I wonder if there is some confusion here about the word "finalizer" in the C# sense and sqlite's use of the term "finalize" for a statement.  Strictly speaking, the C# finalizer for `sqlite3_stmt` should not be in play here.  That gets called only when GC releases the `sqlite3_stmt` object.

Anyway...

Yes, it is true that `raw.sqlite3_finalize` does not protect against being called twice.  This is intentional, because it is trying to have the exact same behavior as the underlying `sqlite_finalize` function.

`sqlite3_stmt` has a Dispose method which calls `sqlite3_finalize`.  It *does* protect against double calls, because it is trying to respect the standard .NET semantics for Dispose().

And `raw.sqlite3_finalize` does try to cooperate with `sqlite3_stmt.Dispose` by setting the flag to make sure it can protect against double calls.

Up at the sqlite-net layer, the stmt class has a Dispose method which also calls `sqlite3_finalize`, but SQLitePCL.raw knows nothing about this.

So, yes, it appears that if you call both Dispose methods (the one in SQLitePCL.raw, and the one in sqlite-net) on the same statement, then you are going to end up calling `sqlite3_finalize` twice, which will crash.  But it seems like it would require a programming error at the app level to do that.

(I suppose sqlite-net could help protect devs from this mistake by *not* calling `sqlite3_finalize` and call `sqlite3_stmt.Dispose` instead.)

It's also possible that the error is in sqlite-net, if it is somehow calling `sqlite3_finalize` twice on its own.  But I haven't seen any indication of that.

Bottom line:  Did you actually see this crash happen?  If so, does your code somehow call Dispose at both layers?

**Comment by ptsoccer** [comment]
Yes I see this crash happening very reliably with my app, though I can't reproduce it on it's own. I have no control over when this insert statement gets disposed as it's cached in SQLiteConnection and I'm using SQLiteAsyncConnection (which internally uses SQLiteConnection eventually).

I think there might be some confusion. I'm not talking about calling dispose directly, I'm literally talking about the c# finalizer. 

If https://github.com/ericsink/SQLitePCL.raw/blob/master/src/cs/intptrs.cs#L217 gets called before https://github.com/praeclarum/sqlite-net/blob/master/src/SQLite.cs#L3026 then this crash will happen. I don't think SQLitePCL has a bug here, I think the issue is sqlite-net should not be calling sqlite3_finalize, and I suspect the reason why it is is because it wasn't always using SQLitePCL and so it had to manage the IntPtr handle on its own and thus needed the finalizer to free that memory.

Anyway, I've spent some time in windbg debugging this and I had made a breakpoint that dumps the clr stack when sqlcipher!sqlite3_finalize is called. Here is that dump:

```
0c7b9a90 free
OS Thread Id: 0x605c (5)
Child SP       IP Call Site
069af4d0 0f5bc9d0 [InlinedCallFrame: 069af4d0] 
069af4c8 0b27bea8 DomainBoundILStubClass.IL_STUB_PInvoke(IntPtr)
069af4d0 0b27bfe7 [InlinedCallFrame: 069af4d0] SQLitePCL.SQLite3Provider_sqlcipher+NativeMethods.sqlite3_finalize(IntPtr)
069af520 0b27bfe7 SQLitePCL.SQLite3Provider_sqlcipher.SQLitePCL.ISQLite3Provider.sqlite3_finalize(IntPtr)
069af524 0b27bfa6 SQLitePCL.raw.sqlite3_finalize(SQLitePCL.sqlite3_stmt)
069af534 0da797df SQLitePCL.sqlite3_stmt.Finalize()
069af730 738363f2 [DebuggerU2MCatchHandlerFrame: 069af730] 

0c7b9a90 free
OS Thread Id: 0x605c (5)
Child SP       IP Call Site
069af470 0f5bc9d0 [InlinedCallFrame: 069af470] 
069af468 0b27bea8 DomainBoundILStubClass.IL_STUB_PInvoke(IntPtr)
069af470 0b27bfe7 [InlinedCallFrame: 069af470] SQLitePCL.SQLite3Provider_sqlcipher+NativeMethods.sqlite3_finalize(IntPtr)
069af4c0 0b27bfe7 SQLitePCL.SQLite3Provider_sqlcipher.SQLitePCL.ISQLite3Provider.sqlite3_finalize(IntPtr)
069af4c4 0b27bfa6 SQLitePCL.raw.sqlite3_finalize(SQLitePCL.sqlite3_stmt)
069af4d4 0b27bf6b SQLite.SQLite3.Finalize(SQLitePCL.sqlite3_stmt) [F:\GitSource\sqlite-net\src\SQLite.cs @ 3534]
069af4e8 0da79927 SQLite.PreparedSqlLiteInsertCommand.Dispose(Boolean) [F:\GitSource\sqlite-net\src\SQLite.cs @ 2678]
069af530 0da7984a SQLite.PreparedSqlLiteInsertCommand.Finalize() [F:\GitSource\sqlite-net\src\SQLite.cs @ 2688]
069af730 738363f2 [DebuggerU2MCatchHandlerFrame: 069af730] 
```

That "0c7b9a90 free" line at the top is the address of the object being freed, notice it's printed twice. The first callstack has the sqlite3_stmt being finalized, the second callstack has the PreparedSqlLiteInsertCommand being finalized.

**Comment by ericsink** [comment]
Hmmm.  Okay, that's interesting.

FWIW, this problem would have first surfaced in 1.1.8, when I added finalizers, at the request of the Entity Framework team.

I want to think more about this.

**Comment by ptsoccer** [comment]
For what it's worth, this seems to reproduce the issue almost every time:
```
using System;
using System.IO;
using SQLite;

namespace SQLiteAccessViolation
{
    class Program
    {
        class TestTable
        {
            public int A { get; set; }
        }
        static void Main(string[] args)
        {

            for (var i = 0; i < 10; ++i)
            {
                var dbName = $"test{i}.db";

                if (File.Exists(dbName))
                    File.Delete(dbName);

                T(dbName);
            }

            GC.Collect();
            GC.WaitForPendingFinalizers();
        }

        static void T(string dbName)
        {
            var conn = new SQLiteConnection(dbName);
            conn.CreateTable<TestTable>();
            conn.Insert(new TestTable() {A = 3});
        }
    }
}
```

Using the latest sqlite net pcl nuget package (1.4.118), I had to update sqlitepclraw to 1.1.8 manually to get the finalizer behavior you mentioned. I experienced this locally since I have a local fork that I recently merged upstream into. I did delete the two finalizers I mentioned earlier and that fixed the problem locally.

**Comment by JKennedy24** [comment]
I am seeing this crash on UWP using Sqlcipher 1.1.9

System.AccessViolationException: 'Attempted to read or write protected memory. This is often an indication that other memory is corrupt.'
   at SQLitePCL.SQLite3Provider_sqlcipher.NativeMethods.sqlite3_step(IntPtr stmt)
   at SQLitePCL.SQLite3Provider_sqlcipher.SQLitePCL.ISQLite3Provider.sqlite3_step(IntPtr stm)
   at SQLite.SQLiteCommand.ExecuteNonQuery()
   at SQLite.SQLiteConnection.Execute(String query, Object[] args)
   at SQLite.SQLiteConnection.Update(Object obj, Type objType)
   at SQLite.SQLiteConnection.Update(Object obj)

Could this be the same error?

**Comment by JKennedy24** [comment]
Mine could potentially be https://github.com/ericsink/SQLitePCL.raw/issues/186 although I can't tell at the moment

**Comment by owndbyu777** [comment]
@JKennedy24  can confirm I am getting that same crash error on UWP importing SQLitePCLRaw.bundle_sqlcipher 1.1.9 through nuget.org and using SQLite.cs in my project. 

Not totally sure what causes it... But it has been thrown when using a "Shared" SQLiteConnection created earlier in the program to insert data, and it has occurred in the Dispose method of an SQLiteConnection.

**Comment by ericsink** [comment]
SQLitePCL.raw pre-release version 1.1.10-pre20180223200113 has been pushed to nuget.  It contains an attempted fix for this problem.  Testing feedback would be appreciated.

**Comment by ptsoccer** [comment]
Tested and it looks good

**Comment by owndbyu777** [comment]
Looks like it has fixed my error up too!

**Comment by UweKeim** [comment]
Just in case anyone will come here via Google, this is my similar stack trace from the Event Log:

```
Frameworkversion: v4.0.30319
Beschreibung: Der Prozess wurde aufgrund einer unbehandelten Ausnahme beendet.
Ausnahmeinformationen: System.AccessViolationException
    bei
SQLitePCL.SQLite3Provider_e_sqlite3+NativeMethods.sqlite3_finalize(IntPtr)
    bei
SQLitePCL.SQLite3Provider_e_sqlite3.SQLitePCL.ISQLite3Provider.sqlite3_finalize(IntPtr)
    bei SQLitePCL.raw.sqlite3_finalize(SQLitePCL.sqlite3_stmt)
    bei SQLite.PreparedSqlLiteInsertCommand.Dispose(Boolean)
    bei SQLite.PreparedSqlLiteInsertCommand.Finalize()
```

**Comment by praeclarum** [comment]
I have bumped raw up to 1.1.11
